### PR TITLE
Server configuration, Stack Arena and Update Makefile

### PR DESCRIPTION
**Pull request by ammarfaizi2**
1. Server configuration (argv parser and config ini parser).
2. Stack arena (allocate huge area of stack for future dynamic memory allocation).
3. Update Makefile (addressing this discussion: https://t.me/TeaInside/34245 and some improvements).

### CODE REVIEW

File: src/sources/teavpn2/global/memory.c [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  __asm__(                          \
+    "mov %0, %%rdi;"                \
+    "mov %1, %%rsi;"                \
+    "mov %2, %%rcx;"                \
+    "rep movsb"                     \
+    :                               \
+    : "r"(DST), "r"(SRC), "r"(N)    \
+    : "rdi", "rsi", "rcx"           \
+  )
+#else
```
Comment (louvian) [inline src/sources/teavpn2/global/memory.c:21]: Can you create unit tests for this macro?
That looks horrible to me.

File: src/sources/teavpn2/server/config.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    RMATCH_N("backlog") {
+      config->sock.backlog = atoi(value);
+    } else {
+      goto invalid_name;
+    }
+
+  } else
+  RMATCH_S("other") {
+
+  } else {
```
Comment (louvian) [inline src/sources/teavpn2/server/config.c:96]: Incomplete config parser?

File: Makefile [hunk 3]
```diff
@@ -4,7 +4,7 @@
 CC     = cc
 CXX    = c++
 NASM   = nasm
```
Comment (louvian) [inline Makefile:6]: Why do we need NASM here?

File: src/sources/teavpn2/server/config.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    RMATCH_N("bind_port") {
+      config->sock.bind_port = (uint16_t)atoi(value);
+    } else
+    RMATCH_N("sock_type") {
+      char targ[4];
+
+      strncpy(targ, value, 3);
+      targ[0] = (targ[0] >= 'A' && targ[0] <= 'Z') ? targ[0] + 32 : targ[0];
+      targ[1] = (targ[1] >= 'A' && targ[1] <= 'Z') ? targ[1] + 32 : targ[1];
+      targ[2] = (targ[2] >= 'A' && targ[2] <= 'Z') ? targ[2] + 32 : targ[2];
```
Comment (louvian) [inline src/sources/teavpn2/server/config.c:73]: Please explain why don't you use `tolower()` from standard library?

**Review by louvian — CHANGES_REQUESTED** [review]
We need to talk in detail :)
I postpone the further review.

File: src/sources/teavpn2/server/config.c (hunk 4 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c:73]: I haven't examined the Assembly, but I think it reduces call overhead by performing those tolower manual arithmetic.

File: Makefile (hunk 3 above)
Comment (ammarfaizi2) [reply Makefile:6]: It is for future support, currently we have no NASM source code.

File: src/sources/teavpn2/server/config.c (hunk 2 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c:96]: Yes, I am still fixing it.

File: src/sources/teavpn2/global/memory.c (hunk 1 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/global/memory.c:21]: Yes, I will be back to you later today.

File: src/sources/teavpn2/global/memory.c [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+
+void *t_ar_alloc(register size_t len)
+{
+  register char *ret  = &(((char *)__t_arena)[__t_arena_pos]);
+  __t_arena_pos      += len;
+
+  return (void *)ret;
+}
```
Comment (zeheater) [inline src/sources/teavpn2/global/memory.c:39]: I don't understand what this function allocate, can you explain more ?

Comment (ammarfaizi2) [reply src/sources/teavpn2/global/memory.c:39]: Yes, it is customized memory allocation. I am going to write the detailed technical about this together with its unit tests soon.

File: src/sources/teavpn2/server/config.c (hunk 2 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c]: Resolved

**Comment by ammarfaizi2** [comment]
@louvian below is the examined situation about `tolower()` consideration.

## Standard tolower() with optimization (-O3)
`tolower()` calls `__ctype_tolower_loc@plt`. 
```
gef➤  r
`/home/ammarfaizi2/project/now/teavpn2/teavpn_server' has changed; re-reading symbols.
Starting program: /home/ammarfaizi2/project/now/teavpn2/teavpn_server ./teavpn_server -c config/server.ini
warning: Probes-based dynamic linker interface failed.
Reverting to original interface.
Parsing arguments...
  config->config_file = "config/server.ini"

Breakpoint 4, server_parser_handler (user=user@entry=0x7fffffffcf60, section=<optimized out>, name=<optimized out>, value=0x7fffffffccfc "tcp", lineno=<optimized out>) at src/sources/teavpn2/server/config.c:75
75	      targ[0] = tolower(targ[0]);
[ Legend: Modified register | Code | Heap | Stack | String ]
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── registers ────
$rax   : 0x00007fffffffcbe4  →  0xfcd0f40000706374 ("tcp"?)
$rbx   : 0x00007fffffffcf60  →  0x00007fffffffe43c  →  "config/server.ini"
$rcx   : 0x00007fffffffcbe7  →  0x8298acfcd0f40000
$rdx   : 0x00007fffffffcbe4  →  0xfcd0f40000706374 ("tcp"?)
$rsp   : 0x00007fffffffcbe0  →  0x00706374f7f5b3c0
$rbp   : 0x00007fffffffccfc  →  0x0000000000706374 ("tcp"?)
$rsi   : 0x00007fffffffccfc  →  0x0000000000706374 ("tcp"?)
$rdi   : 0x00007fffffffcbe4  →  0xfcd0f40000706374 ("tcp"?)
$rip   : 0x0000555555555c35  →  <server_parser_handler+757> call 0x555555555180 <__ctype_tolower_loc@plt>
$r8    : 0x0               
$r9    : 0x00007fffffffcc70  →  0x000074656b636f73 ("socket"?)
$r10   : 0x00007ffff7f5aac0  →  0x0000000100000000
$r11   : 0x7079745f6b636f73 ("sock_typ"?)
$r12   : 0x0               
$r13   : 0x00007fffffffcbe4  →  0xfcd0f40000706374 ("tcp"?)
$r14   : 0x00007fffffffccf0  →  "sock_type"
$r15   : 0x00007ffff7f5b3c0  →  0x0002000200020002
$eflags: [ZERO carry PARITY adjust sign trap INTERRUPT direction overflow resume virtualx86 identification]
$cs: 0x0033 $ss: 0x002b $ds: 0x0000 $es: 0x0000 $fs: 0x0000 $gs: 0x0000 
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── stack ────
0x00007fffffffcbe0│+0x0000: 0x00706374f7f5b3c0	 ← $rsp
0x00007fffffffcbe8│+0x0008: 0x778298acfcd0f400
0x00007fffffffcbf0│+0x0010: 0x00007fffffffd310  →  0x0000000000000000
0x00007fffffffcbf8│+0x0018: 0x00007fffffffccfe  →  0x2e00000000000070 ("p"?)
0x00007fffffffcc00│+0x0020: 0x00007fffffffcf20  →  0x000055555555a6b0  →  0x00000000fbad2488
0x00007fffffffcc08│+0x0028: 0x000000000000000b
0x00007fffffffcc10│+0x0030: 0x00007fffffffccf0  →  "sock_type"
0x00007fffffffcc18│+0x0038: 0x0000555555555f61  →  <ini_parse_stream+481> mov ecx, DWORD PTR [rsp+0x3c]
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x555555555c2a <server_parser_handler+746> mov    rsi, rbp
   0x555555555c2d <server_parser_handler+749> mov    rdi, r13
   0x555555555c30 <server_parser_handler+752> call   0x5555555550f0 <strncpy@plt>
 → 0x555555555c35 <server_parser_handler+757> call   0x555555555180 <__ctype_tolower_loc@plt>
   ↳  0x555555555180 <__ctype_tolower_loc@plt+0> endbr64 
      0x555555555184 <__ctype_tolower_loc@plt+4> bnd    jmp QWORD PTR [rip+0x3e25]        # 0x555555558fb0 <__ctype_tolower_loc@got.plt>
      0x55555555518b <__ctype_tolower_loc@plt+11> nop    DWORD PTR [rax+rax*1+0x0]
      0x555555555190 <__ctype_b_loc@plt+0> endbr64 
      0x555555555194 <__ctype_b_loc@plt+4> bnd    jmp QWORD PTR [rip+0x3e1d]        # 0x555555558fb8 <__ctype_b_loc@got.plt>
      0x55555555519b <__ctype_b_loc@plt+11> nop    DWORD PTR [rax+rax*1+0x0]
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── arguments (guessed) ────
__ctype_tolower_loc@plt (
)
──────────────────────────────────────────────────────────────────────────────────────────────────────────────── source:src/sources/tea[...].c+75 ────
     70	 
     71	       strncpy(targ, value, 3);
     72	       // targ[0] = (targ[0] >= 'A' && targ[0] <= 'Z') ? targ[0] + 32 : targ[0];
     73	       // targ[1] = (targ[1] >= 'A' && targ[1] <= 'Z') ? targ[1] + 32 : targ[1];
     74	       // targ[2] = (targ[2] >= 'A' && targ[2] <= 'Z') ? targ[2] + 32 : targ[2];
               // targ=0x00007fffffffcbe4  →  0xfcd0f40000706374 ("tcp"?)
 →   75	       targ[0] = tolower(targ[0]);
     76	       targ[1] = tolower(targ[1]);
     77	       targ[2] = tolower(targ[2]);
     78	       targ[3] = '\0';
     79	 
     80	       if (!strcmp(targ, "tcp")) {
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "teavpn_server", stopped, reason: BREAKPOINT
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x555555555c35 → server_parser_handler(user=0x7fffffffcf60, section=<optimized out>, name=<optimized out>, value=0x7fffffffccfc "tcp", lineno=<optimized out>)
[#1] 0x555555555f61 → ini_parse_stream(reader=0x7ffff7e417b0, stream=0x55555555a6b0, handler=0x555555555940 <server_parser_handler>, user=0x7fffffffcf60)
[#2] 0x5555555563b9 → ini_parse_file(file=0x55555555a6b0, handler=0x555555555940 <server_parser_handler>, user=0x7fffffffcf60)
[#3] 0x5555555563f1 → ini_parse(filename=<optimized out>, handler=0x555555555940 <server_parser_handler>, user=0x7fffffffcf60)
[#4] 0x555555555cf7 → tvpn_server_load_config_file(file=<optimized out>, config=0x7fffffffcf60)
[#5] 0x555555555223 → main(argc=0x4, argv=0x7fffffffe0d8, envp=0x7fffffffe100)
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤ 
```

## Customized tolower() with optimization (-O3)
We have no call to function to do tolower().
```
gef➤  r
Starting program: /home/ammarfaizi2/project/now/teavpn2/teavpn_server -c config/server.ini
Parsing arguments...
  config->config_file = "config/server.ini"

Breakpoint 2, server_parser_handler (user=user@entry=0x7fffffffcf70, section=<optimized out>, name=<optimized out>, value=0x7fffffffcd0c "tcp", lineno=<optimized out>) at src/sources/teavpn2/server/config.c:72
72	      targ[0] = (targ[0] >= 'A' && targ[0] <= 'Z') ? targ[0] + 32 : targ[0];
[ Legend: Modified register | Code | Heap | Stack | String ]
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── registers ────
$rax   : 0x00007fffffffcc04  →  0x1eaadd0000706374 ("tcp"?)
$rbx   : 0x00007fffffffcf70  →  0x00007fffffffe43c  →  "config/server.ini"
$rcx   : 0x00007fffffffcc07  →  0x8d38211eaadd0000
$rdx   : 0x00007fffffffcc04  →  0x1eaadd0000706374 ("tcp"?)
$rsp   : 0x00007fffffffcc00  →  0x00706374ffffd320
$rbp   : 0x00007fffffffcd0c  →  0x0000000000706374 ("tcp"?)
$rsi   : 0x00007fffffffcd0c  →  0x0000000000706374 ("tcp"?)
$rdi   : 0x0000555555557004  →  0x0070647500706374 ("tcp"?)
$rip   : 0x0000555555555c14  →  <server_parser_handler+756> movzx eax, BYTE PTR [rsp+0x4]
$r8    : 0x00007fffffffcc04  →  0x1eaadd0000706374 ("tcp"?)
$r9    : 0x00007fffffffcc80  →  0x000074656b636f73 ("socket"?)
$r10   : 0x00007ffff7f5aac0  →  0x0000000100000000
$r11   : 0x7079745f6b636f73 ("sock_typ"?)
$r12   : 0x0               
$r13   : 0x00007fffffffcd00  →  "sock_type"
$r14   : 0x00007fffffffcd00  →  "sock_type"
$r15   : 0x00007ffff7f5b3c0  →  0x0002000200020002
$eflags: [ZERO carry PARITY adjust sign trap INTERRUPT direction overflow resume virtualx86 identification]
$cs: 0x0033 $ss: 0x002b $ds: 0x0000 $es: 0x0000 $fs: 0x0000 $gs: 0x0000 
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── stack ────
0x00007fffffffcc00│+0x0000: 0x00706374ffffd320	 ← $rsp
0x00007fffffffcc08│+0x0008: 0xb78d38211eaadd00
0x00007fffffffcc10│+0x0010: 0x00007fffffffcd0e  →  0x2e00000000000070 ("p"?)
0x00007fffffffcc18│+0x0018: 0x00007fffffffcf30  →  0x000055555555a6b0  →  0x00000000fbad2488
0x00007fffffffcc20│+0x0020: 0x000000000000000b
0x00007fffffffcc28│+0x0028: 0x0000555555555f41  →  <ini_parse_stream+481> mov ecx, DWORD PTR [rsp+0x3c]
0x00007fffffffcc30│+0x0030: 0x0000000000000000
0x00007fffffffcc38│+0x0038: 0x00007fffffffcd0a  →  0x0000007063742000
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x555555555c05 <server_parser_handler+741> mov    BYTE PTR [rsp+0x7], 0x0
   0x555555555c0a <server_parser_handler+746> lea    rdi, [rip+0x13f3]        # 0x555555557004
   0x555555555c11 <server_parser_handler+753> mov    r8, rax
 → 0x555555555c14 <server_parser_handler+756> movzx  eax, BYTE PTR [rsp+0x4]
   0x555555555c19 <server_parser_handler+761> mov    rsi, r8
   0x555555555c1c <server_parser_handler+764> lea    ecx, [rax-0x41]
   0x555555555c1f <server_parser_handler+767> lea    edx, [rax+0x20]
   0x555555555c22 <server_parser_handler+770> cmp    cl, 0x1a
   0x555555555c25 <server_parser_handler+773> cmovb  eax, edx
──────────────────────────────────────────────────────────────────────────────────────────────────────────────── source:src/sources/tea[...].c+72 ────
     67	     } else
     68	     RMATCH_N("sock_type") {
     69	       char targ[4];
     70	 
     71	       strncpy(targ, value, 3);
               // targ=0x00007fffffffcc04  →  0x1eaadd0000706374 ("tcp"?)
 →   72	       targ[0] = (targ[0] >= 'A' && targ[0] <= 'Z') ? targ[0] + 32 : targ[0];
     73	       targ[1] = (targ[1] >= 'A' && targ[1] <= 'Z') ? targ[1] + 32 : targ[1];
     74	       targ[2] = (targ[2] >= 'A' && targ[2] <= 'Z') ? targ[2] + 32 : targ[2];
     75	       // targ[0] = tolower(targ[0]);
     76	       // targ[1] = tolower(targ[1]);
     77	       // targ[2] = tolower(targ[2]);
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "teavpn_server", stopped, reason: BREAKPOINT
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x555555555c14 → server_parser_handler(user=0x7fffffffcf70, section=<optimized out>, name=<optimized out>, value=0x7fffffffcd0c "tcp", lineno=<optimized out>)
[#1] 0x555555555f41 → ini_parse_stream(reader=0x7ffff7e417b0 <_IO_fgets>, stream=0x55555555a6b0, handler=0x555555555920 <server_parser_handler>, user=0x7fffffffcf70)
[#2] 0x555555556399 → ini_parse_file(file=0x55555555a6b0, handler=0x555555555920 <server_parser_handler>, user=0x7fffffffcf70)
[#3] 0x5555555563d1 → ini_parse(filename=<optimized out>, handler=0x555555555920 <server_parser_handler>, user=0x7fffffffcf70)
[#4] 0x555555555cd7 → tvpn_server_load_config_file(file=<optimized out>, config=<optimized out>)
[#5] 0x555555555203 → main(argc=0x3, argv=0x7fffffffe0e8, envp=0x7fffffffe108)
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  
```

File: src/sources/teavpn2/global/memory.c (hunk 1 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/global/memory.c]: The unit tests have been created https://github.com/TeaInside/teavpn2/pull/8/commits/d87f66acbda58ede0d88540b866feeb7213f0372

File: src/sources/teavpn2/server/config.c (hunk 4 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c:73]: Can we resolve this tolower problem now?

**Comment by louvian** [comment]
I think there is nothing wrong with this `tolower()`, let's go for it!

**Comment by louvian** [comment]
@ammarfaizi2 is this ready for re-review?

**Comment by Sunda001** [comment]
@louvian interesting

**Comment by ammarfaizi2** [comment]
@louvian Yes, we it seems it is ready for re-review.
Cc: @zeheater

**Comment by ammarfaizi2** [comment]
@louvian I am going to add CI in the near future before this PR be merged. So please don't merge it in case you approve the latest changes.

Cc: @zeheater

File: Makefile [hunk 6]
```diff
@@ -16,30 +16,38 @@ GLOBAL_SOURCE_DIR = $(SOURCES_DIR)/teavpn2/global
 CLIENT_SOURCE_DIR = $(SOURCES_DIR)/teavpn2/client
 SERVER_SOURCE_DIR = $(SOURCES_DIR)/teavpn2/server
 
-# C/C++ compile flag.
-CC_COMPILE_FLAGS  = -std=c99 $(INCLUDE_DIR) -c
-CXX_COMPILE_FLAGS = -std=c++17 $(INCLUDE_DIR) -D_GLIBCXX_ASSERTIONS -c
+DEFAULT_OPTIMIZATION = -O3
```
Comment (louvian) [inline Makefile:19]: Are you sure `-O3` as default optimization?

File: src/sources/teavpn2/global/third_party/inih/inih.c [hunk 7]
```diff
@@ -72,12 +72,19 @@ static char* find_chars_or_comment(const char* s, const char* chars)
 }
 
 /* Version of strncpy that ensures dest (size bytes) is null-terminated. */
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+#endif
```
Comment (louvian) [inline src/sources/teavpn2/global/third_party/inih/inih.c:78]: Is this a warning supression?

File: src/sources/teavpn2/server/config.c (hunk 4 above)
Comment (louvian) [reply src/sources/teavpn2/server/config.c:74]: Resolved

File: tests/test.mk [hunk 8]
```diff
@@ -0,0 +1,74 @@
+
+
+CRITERION_DIR      = criterion-v2.3.3
+CRITERION_TAR      = criterion-v2.3.3-linux-x86_64.tar.bz2
+
+INCLUDE_DIR       := $(INCLUDE_DIR:-I%=-I../%) -I$(CRITERION_DIR)/include
+SOURCES_DIR        = sources
+BIN_DIR			   = .bin
```
Comment (louvian) [inline tests/test.mk:8]: Too many spaces or tabs, keep them inline so it looks better.

File: src/sources/teavpn2/server/config.c [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      // targ[0] = tolower(targ[0]);
+      // targ[1] = tolower(targ[1]);
+      // targ[2] = tolower(targ[2]);
```
Comment (louvian) [inline src/sources/teavpn2/server/config.c:77]: You can remove this

File: src/sources/teavpn2/server/argv_parser.c [hunk 10]
```diff
@@ -106,6 +141,97 @@ inline static bool getopt_handler(int argc, char **argv, server_config *config)
...
+  printf("\n");
+  printf("Available options:\n");
+  printf("  -h, --help\t\t\tShow this help message.\n");
+  printf("  -c, --config=FILE\t\tSet config file.\n");
+
+  printf("\n");
+  printf("[Config options]\n");
+  printf("Virtual network interface:\n");
+  printf("  -d, --dev=DEV\t\t\tSet virtual network interface name (default: %s).\n", default_dev_name);
+  printf("  -m, --mtu=MTU\t\t\tSet mtu value (default: %d).\n", default_mtu);
```
Comment (louvian) [inline src/sources/teavpn2/server/argv_parser.c:216]: Maybe add `--version` option?

File: src/sources/teavpn2/server/config.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+
+bool tvpn_server_load_config_file(char *file, server_cfg *config)
+{
+  if (ini_parse(file, server_parser_handler, config) < 0) {
+    return false;
+  }
+
+  return true;
+}
```
Comment (louvian) [inline src/sources/teavpn2/server/config.c:26]: It can be simplified in one line

File: tests/Makefile [hunk 12]
```diff
@@ -0,0 +1,33 @@
+
+ifndef TARGET_TEST
+$(error TARGET_TEST is not set)
+endif
+
+SOURCES   = $(shell find ${TARGET_TEST} -name '*.c')
```
Comment (louvian) [inline tests/Makefile:6]: No C++ support for unit tests?

File: .github/workflows/c-cpp.yml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      run: make test
+
+    ########################################
+    #      Store generated artifacts       #
+    ########################################
+    - name: Store artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: teavpn_server
+        path: teavpn_server
```
Comment (louvian) [inline .github/workflows/c-cpp.yml:65]: Why don't you store `teavpn_client` too?

File: .github/workflows/c-cpp.yml [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+          # Note: 4.0.1 has no 16.04 package and the OSX archive extracts
+          #  to a different arch, so we don't use it
+          - { target: clang-4.0.0, compiler: clang, cxx-version: 4.0.0 }
+          - { target: clang-3.9.0, compiler: clang, cxx-version: 3.9.0 }
+          # g++ boilerplace
+          - { target: g++-9, compiler: g++, cxx-version: 9.3.0 }
+          - { target: g++-8, compiler: g++, cxx-version: 8.4.0 }
+          - { target: g++-7, compiler: g++, cxx-version: 7.5.0 }
+          - { target: g++-6, compiler: g++, cxx-version: 6.5.0 }
+          - { target: g++-5, compiler: g++, cxx-version: 5.5.0 }
```
Comment (louvian) [inline .github/workflows/c-cpp.yml:46]: Isn't it overkill to support these many compilers?
I would like to build on different machine instead different compiler in the same machine.
However, we can take severals gcc and clang in the same machine, but don't be this many.

File: Makefile (hunk 3 above)
Comment (louvian) [reply Makefile:6]: I don't know why we need NASM when GAS exists.

**Review by louvian — CHANGES_REQUESTED** [review]

File: Makefile (hunk 6 above)
Comment (ammarfaizi2) [reply Makefile]: Resolved, changed to `-O0`.

File: src/sources/teavpn2/global/third_party/inih/inih.c (hunk 7 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/global/third_party/inih/inih.c:78]: Yes, it is.

File: tests/test.mk (hunk 8 above)
Comment (ammarfaizi2) [reply tests/test.mk]: Resolved.

File: src/sources/teavpn2/server/argv_parser.c (hunk 10 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/argv_parser.c:216]: Not really important for now.

File: src/sources/teavpn2/server/config.c (hunk 9 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c:77]: Resolved.

File: tests/Makefile (hunk 12 above)
Comment (ammarfaizi2) [reply tests/Makefile:6]: Currently, no need to add C++ support for unit tests.

File: src/sources/teavpn2/server/config.c (hunk 11 above)
Comment (ammarfaizi2) [reply src/sources/teavpn2/server/config.c:26]: Resolved.

File: .github/workflows/c-cpp.yml (hunk 13 above)
Comment (ammarfaizi2) [reply .github/workflows/c-cpp.yml]: Resolved.

File: .github/workflows/c-cpp.yml (hunk 14 above)
Comment (ammarfaizi2) [reply .github/workflows/c-cpp.yml]: Resolved.

File: Makefile (hunk 3 above)
Comment (ammarfaizi2) [reply Makefile:6]: We can discuss it later. It think it is not really important for now.

**Comment by ammarfaizi2** [comment]
Requested changes are all resolved, please re-review @louvian.

**Comment by ammarfaizi2** [comment]
Forgot to push the latest changes, should be ok now.

**Review by zeheater — APPROVED** [review]
Looks good

**Review by louvian — APPROVED** [review]
Good, ready to merge :)

**Comment by louvian** [comment]
Note before merge @ammarfaizi2:
We will have to make the internal documentation that explains the custom mechanism like allocator and memcpy in the above commits (in the next PR maybe), hope it can be achieved while hacktoberfest is still going.

**Comment by ammarfaizi2** [comment]
@louvian yes, I am responsible for that work. Will do that soon.

**Comment by ammarfaizi2** [comment]
Merged, thanks for the review.

**Pull request merged**